### ISSUE TITLE
Drop support for Java 5 runtime?

### ISSUE THREAD

**Issue by marchof** [jacoco/jacoco#314]
I wonder whether we could drop support for Java 5 runtime with the following benefits:

* Simplify test/build setup: With the advent of Java 9 we're now testing on 5 different JDK versions.
* Allow dynamic attachment of JaCoCo agent
* Simplify runtime data access
* allow coverage on bootstrap classes
* support re-transformation (no schema changes any more)

The last three become possible by adding companion classes to the bootstrap classpath with new method java.lang.instrument.Instrumentation.appendToBootstrapClassLoaderSearch().

**Comment by kibertoad** [comment]
So much yes.

**Comment by Godin** [comment]
@marchof I'm not against drop of support.

However to prevent misunderstanding for others:
* we already allow coverage on bootstrap classes
  * in online instrumentation mode partially due to inability to retransform already loaded classes
  * and I suppose that completely in offline mode?
* ability to use "appendToBootstrapClassLoaderSearch" for companion classes still requires practical confirmation
* might be other ways for bootstrap classes - subject needs to be studied more deeply?
* I believe that dynamic attachment of agent might be added without removal of support of Java 5

But indeed - removal of support simplifies things.

**Comment by silviu-burcea** [comment]
Oracle is planning to support only the last 3(current + previous 2)
versions of Java. You may want to follow this policy too.

On Fri, May 22, 2015 at 11:48 PM, Evgeny Mandrikov <notifications@github.com
> wrote:

> @marchof <https://github.com/marchof> I'm not against drop of support.
>
> However to prevent misunderstanding for others:
>
>    - we already allow coverage on bootstrap classes
>       - in online instrumentation mode partially due to inability to
>       retransform already loaded classes
>       - and I suppose that completely in offline mode?
>     - ability to use "appendToBootstrapClassLoaderSearch" for companion
>    classes still requires practical confirmation
>    - might be other ways for bootstrap classes - subject needs to be
>    studied more deeply?
>    - I believe that dynamic attachment of agent might be added without
>    removal of support of Java 5
>
> But indeed - removal of support simplifies things.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jacoco/jacoco/issues/314#issuecomment-104763758>.
>



-- 
- OCPJP7 (90%)
- OCAJP7 (93%)
- Java and Big Data Enthusiast

**Comment by marchof** [comment]
@Godin Of course, first we need to verify that the new approach to share probe instances actually works. I just wanted to open this as a general discussion whether Java 5 runtime is still required.

**Comment by silviuburceadev** [comment]
I can't believe I've commented on this thread 10 years ago, but I'd like to suggest you to reconsider dropping support for older versions. Sure, you can brag that you can have code coverage for even Java 1 (according to FAQ), but let's be a bit practical here, the market share of even Java 7 is non-existent at this point, with 8, 11 and 17 dominating and 21 slowly ramping up. It makes 0 sense to keep supporting Java 5 and limit yourself from having better implementations and newer toolchain, e.g. in org.jacoco.build module:

```
<plugin>
          <groupId>org.apache.maven.plugins</groupId>
          <artifactId>maven-surefire-plugin</artifactId>
          <!--  latest version which runs with Java 5 -->
          <version>2.19.1</version>
        </plugin>
```

**Comment by marchof** [comment]
@silviuburceadev Thanks for checking back after long time. We have to carefully distinguish two Java versions here:

1) Class files versions under test: Yes, we support all class file versions since the beginning. I still see this this as a strong requirement as projects depend on third-party libs that for whatever reason compile to older Java versions. For example many of the apache-commons libraries are still compiled to Java 5. If we would require Java 8 for the class files we would render JaCoCo unusable for most projects dues to transitive dependencies. Beside testing there is no real issue with this kind of backward compatibility in JaCoCo as class files specification was and is very stable.
2) Test runtime Java version: As you noticed here Java 5 compatibility gives us some limitations like library dependencies, build tools and also language/API features in our own code. But until now we didn't hit a real blocker, so I prefer to keep JaCoCo usable for the many legacy projects out there that depend on JaCoCo for testing. But sure, if we will hit a blocker we will definitely consider giving up older Java versions in favor of future-proof maintenance of JaCoCo
